nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp

```diff
@@ -0,0 +1,880 @@
+
+      // process first column. Interpret pixel from column -1 as 0
+      {
+        auto overlay = {uint8_t(0), apply_shape(*in, 1), apply_shape(*(in + 1), 2)};
+        touch_fn(*out, overlay);
+        ++in;
+        ++out;
+      }
+
+      // process next columns up to last
+      for (; in != last_column_pixel; ++in, ++out) {
+        auto overlay = {
+          apply_shape(*(in - 1), 0),
+          apply_shape(*(in), 1),
+          apply_shape(*(in + 1), 2)
+        };
+        touch_fn(*out, overlay);
+      }
+
+      // process last column
+      {
+        auto overlay = {apply_shape(*(in - 1), 0), apply_shape(*(in), 1), uint8_t(0)};
+        touch_fn(*out, overlay);
+        ++in;
+        ++out;
+      }
+    }
+  }
+}
+
+/**
+ * @brief Perform morphological operations
+ * @tparam AggregateFn function object.
+ * Signature should be equivalent to the following:
+ * uint8_t fn(std::initializer_list<uint8_t> in),
+ * where in - result of overlaying the shape on the neighborhood of source pixel
+ * @param input input image
+ * @param output output image
+ * @param shape structuring element image with size 3x3.
+ * Should only contain values 0 (ignore neighborhood pixel) or 255 (use pixel).
+ * @param aggregate neighborhood pixels aggregator
+ * @throw std::logic_error if the sizes of the input and output images are different or
+ * shape size is not equal to 3x3
+ */
+template<class AggregateFn>
+void morphology_operation(
+  const Image<uint8_t> & input, Image<uint8_t> & output,
+  const Image<uint8_t> & shape, AggregateFn aggregate)
+{
+  if (input.rows() != output.rows() || input.columns() != output.columns()) {
+    throw std::logic_error(
+            "morphology_operation: the sizes of the input and output images are different");
+  }
+
+  if (shape.rows() != 3 || shape.columns() != 3) {
+    throw std::logic_error("morphology_operation: wrong shape size");
+  }
+
+  if (input.empty()) {
+    return;
+  }
+
+  // Simple write the pixel of the output image (first pass only)
+  auto set = [&](uint8_t & res, std::initializer_list<uint8_t> lst) {res = aggregate(lst);};
+  // Update the pixel of the output image
+  auto update = [&](uint8_t & res, std::initializer_list<uint8_t> lst) {
+      res = aggregate({res, aggregate(lst)});
+    };
+
+  // Apply the second shape row
+  probeRows(input.row(0), output.row(0), input.rows(), input.columns(), shape.row(1), set);
+
+  if (input.rows() > 1) {
+    // Apply the first (top) shape row
+    probeRows(input.row(0), output.row(1), input.rows() - 1, input.columns(), shape.row(0), update);
+    // Apply the last shape row
+    probeRows(input.row(1), output.row(0), input.rows() - 1, input.columns(), shape.row(2), update);
+  }
+}
+
+/// @brief Return structuring element 3x3 image by predefined figure type
+Image<uint8_t> getShape(ConnectivityType connectivity)
+{
+  Image<uint8_t> shape(3, 3);
+
+  if (connectivity == ConnectivityType::Way8) {
+    shape.fill(255);
```

Comment:
The values `0` and `255` in the figure are the flags to use/not use the corresponding pixel when convolving. I pointed this out in [this comment](https://github.com/ryzhikovas/navigation2-fork/blob/feature-costmap2d-denoise/nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp#L813). If I have to move away from magic constants, then it would be better to move to the bool type and `true`/`false` values. But `0`/`255` seems to be conventional for masks...

Reply:
I am OK about `0/255` values, there is no need to change them for `true/false`. Could you please make the constants to express the essence instead of hard-coding `0` and `255`. Say:
```
> nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp
...
static constexpr uint8_t IGNORE_PIXEL = 0;
static constexpr uint8_t USE_PIXEL = 255;
```
Or something similar?

Reply:
[Introduced](https://github.com/ros-planning/navigation2/pull/2567/commits/a8e657d4706ee747ee14b45685e51446011aa3c2)  single-letter variables with an explanation of their meaning in the comment. Due to this, it was possible to arrange the matrices in a human-readable form. I think given the very short lifetime, short variables are acceptable.

Reply:
The last change looks good. Closing item.